src/daemon.rs

```diff
@@ -138,3 +145,305 @@ impl fmt::Display for NotifyState {
+}
+
+/// Check for file descriptors passed by systemd
+///
+/// Like sd_listen_fds, but will also return a Vec of names associated with each file
+/// descriptor.
+pub fn sd_listen_fds_with_names(unset_env: bool) -> Result<Vec<(SocketType, String)>> {
+    let pid = env::var("LISTEN_PID")?;
+    let fds = env::var("LISTEN_FDS")?;
+    let names = env::var("LISTEN_FDNAMES")?;
+
+    if unset_env {
+        env::remove_var("LISTEN_PID");
+        env::remove_var("LISTEN_FDS");
+        env::remove_var("LISTEN_FDNAMES");
+    }
+
+    let pid = pid.parse::<u32>()?;
+    let fds = fds.parse::<i32>()?;
+
+    if process::id() != pid {
+        return Err("Pid mismatch".into());
+    }
+
+    let names: Vec<String> = names.split(":").map(String::from).collect();
+    let vec = socks_from_fds(fds);
+    let out = vec.into_iter().zip(names.into_iter()).collect();
+
+    Ok(out)
+}
+
+fn socks_from_fds(fds: RawFd) -> Vec<SocketType> {
+    let mut vec = Vec::new();
+    for fd in SD_LISTEN_FDS_START..SD_LISTEN_FDS_START + fds {
+        if let Ok(sock) = SocketType::try_from(fd) {
+            vec.push(sock);
+        } else {
+            eprintln!("Socket conversion error");
+        }
+    }
+
+    vec
+}
+
+impl IsType for RawFd {
+    fn is_fifo(&self) -> bool {
+        match fstat(*self) {
+            Ok(stat) => (stat.st_mode & 0170000) == 0010000,
+            Err(_) => false,
+        }
+    }
+
+    fn is_special(&self) -> bool {
+        match fstat(*self) {
+            Ok(stat) => (stat.st_mode & 0170000) == 0100000,
+            Err(_) => false,
+        }
+    }
+
+    fn is_inet(&self) -> bool {
+        match getsockname(*self) {
+            Ok(addr) => {
+                if let SockAddr::Inet(_) = addr {
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+            Err(_) => return false,
+        }
+    }
+
+    fn is_unix(&self) -> bool {
+        match getsockname(*self) {
+            Ok(addr) => {
+                if let SockAddr::Unix(_) = addr {
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+            Err(_) => return false,
+        }
+    }
+
+    fn is_mq(&self) -> bool {
+        mq_getattr(*self).is_ok()
+    }
+}
+
+impl TryFrom<RawFd> for SocketType {
+    type Error = &'static str;
+
+    fn try_from(value: RawFd) -> std::result::Result<Self, Self::Error> {
+        if value.is_fifo() {
+            return Ok(SocketType::Fifo(value));
+        } else if value.is_special() {
+            return Ok(SocketType::Special(value));
+        } else if value.is_inet() {
+            return Ok(SocketType::Inet(value));
+        } else if value.is_unix() {
+            return Ok(SocketType::Unix(value));
+        } else if value.is_mq() {
+            return Ok(SocketType::Mq(value));
+        }
+
+        return Err("Invalid FD");
+    }
+}
+
+#[cfg(test)]
```

Comment:
These tests have a few issues:
 * they assume an external tool is present (`ncat`)
 * they assume `127.0.0.1:7878` is not already in use
 * the manipulate the environment, which is shared across threads, and can thus race

I suggest we drop them completely and just add some basic unit tests for the things that can be safely tested (e.g. the `is_*`). We can come back and add better high-level ones later on.

Reply:
That works for me, although I'm not sure how else to exercise the sd_listen* functions without some side effects

Reply:
It is fine too if we don't go for full coverage here. Integration tests relying on OS global resources are tricky, and the flake/utility ratio is not so high.